### DISCUSSION TITLE
A Tour and Demo of Wave doesn't work (VPN/Proxy related)

### DISCUSSION THREAD

**Discussion by freegor** [h2oai/wave#725] [Q&A]
Hi there,
I have installed the Wave 0.13.0 20210306054523 on my Mac OS Catalina.
I have installed Python  3.9.2.
After the installation the Wave is up, I see in bash:

_2021/04/07 19:26:38 # ┌─────────────────────────┐
2021/04/07 19:26:38 # │  ┌    ┌ ┌──┐ ┌  ┌ ┌──┐  │ H2O Wave
2021/04/07 19:26:38 # │  │ ┌──┘ │──│ │  │ └┐    │ 0.13.0 20210306054523
2021/04/07 19:26:38 # │  └─┘    ┘  ┘ └──┘  └─┘  │ © 2021 H2O.ai, Inc.
2021/04/07 19:26:38 # └─────────────────────────┘
2021/04/07 19:26:38 # 
2021/04/07 19:26:38 # {"address":":10101","t":"listen","webroot":"/Users/user/wave/wave-0.13.0-darwin-amd64/www"}_

and I can see the "WAVE LIVE" banner in the browser.
Once I start the Tour, everything looks good in the Terminal tab:

(venv) wave-0.13.0-darwin-amd64 user$ wave run --no-reload examples.tour
 _Welcome to the H2O Wave Interactive Tour!
 Go to http://localhost:10101/tour
INFO:     Started server process [23623]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)_

but in the browser on the _/tour_
I see the same blinking "WAVE LIVE" banner as well as on the _/demo_
Some ideas?

**Comment by tkmaker** [comment]
hi @freegor What messages do you see in your running wave server once you start the example tour? 

This is an example of what you should expect to see once you have started the tour and then navigate to: `http://localhost:10101/tour` in a browser

```
2021/04/07 13:15:44 # {"address":":10101","t":"listen","webroot":"/Users/tkalyanpur/Desktop/wave/wave-0.13.0-darwin-amd64/www"}
2021/04/07 13:15:53 # {"host":"http://127.0.0.1:8000","route":"/tour","t":"app_add"}
2021/04/07 13:17:31 # {"addr":"[::1]:55276","route":"/","t":"ui_add"}
2021/04/07 13:17:35 # {"addr":"[::1]:55276","t":"ui_drop"}
2021/04/07 13:17:35 # {"addr":"[::1]:55277","route":"/tour","t":"ui_add"}
2021/04/07 13:17:35 # {"addr":"[::1]:55277","route":"/976ef3f7-fa9d-4a56-a986-cb4bddd70626","t":"ui_add"}
2021/04/07 13:17:35 * /976ef3f7-fa9d-4a56-a986-cb4bddd70626 {"d":[{"k":"meta","d":{"view":"meta","box":"","title":"H2O Wave Tour"}},{"k":"header","d":{"view":"header","box":"1 1 2 1","title":"H2O Wave Tour","subtitle":"178 Interactive Examples"}},{"k":"examples","d":{"view":"nav","box":"1 2 2 9","items":[{"label":"Examples","items":[{"name":"#hello_world","label":"Hello World!"},{"name":"#todo","label":"To-do List App"},{"name":"#wizard","label":"Wizard"},{"name":"#issue_tracker","label":"Issue Tracker"},{"name":"#dashboard","label":"Dashboard"},{"name":"#stat_small","label":"Stat / Small"},{"name":"#stat_large","label":"Stat / Large"},{"name":"#stat_wide_gauge","label":"Stat / Gauge / Wide"},{"name":"#stat_tall_gauge","label":"Stat / Gauge / Tall"},{"name":"#stat_wide_bar","label":"Stat / Bar / Wide"},{"name":"#stat_large_bar","label":"Stat / Bar / Large"},{"name":"#stat_small_series_area","label":"Stat / Series / Small / Area"},{"name":"#stat_small_series_interval","label":"Stat / Series / Small / Interval"},{"name":"#stat_wide_series_area","label":"Stat / Series / Wide / Area"},{"name":"#stat_wide_series_interval","label":"Stat / Series / Wide / Interval"},{"name":"#stat_tall_series_area","label":"Stat / Series / Tall / Area"},{"name":"#stat_tall_series_interval","label":"Stat / Series / Tall / Interval"},{"name":"#layout","label":"Layout / Position"},{"name":"#layout_size","label":"Layout / Size"},{"name":"#layout_responsive","label":"Layout / Responsive"},{"name":"#form","label":"Form"},{"name":"#text","label":"Form / Text"},{"name":"#text_sizes","label":"Form / Text / Sizes"},{"name":"#label","label":"Form / Label"},{"name":"#link","label":"Form / Link"},{"name":"#progress....
```

**Reply by freegor** [reply]
wave-0.13.0-darwin-amd64 user$ ./waved 
2021/04/07 19:26:38 # 
2021/04/07 19:26:38 # ┌─────────────────────────┐
2021/04/07 19:26:38 # │  ┌    ┌ ┌──┐ ┌  ┌ ┌──┐  │ H2O Wave
2021/04/07 19:26:38 # │  │ ┌──┘ │──│ │  │ └┐    │ 0.13.0 20210306054523
2021/04/07 19:26:38 # │  └─┘    ┘  ┘ └──┘  └─┘  │ © 2021 H2O.ai, Inc.
2021/04/07 19:26:38 # └─────────────────────────┘
2021/04/07 19:26:38 # 
2021/04/07 19:26:38 # {"address":":10101","t":"listen","webroot":"/Users/user/wave/wave-0.13.0-darwin-amd64/www"}
2021/04/07 19:34:25 # {"addr":"127.0.0.1:58275","route":"/","t":"ui_add"}
2021/04/07 19:34:27 # {"addr":"127.0.0.1:58275","t":"ui_drop"}
2021/04/07 19:37:19 # {"addr":"127.0.0.1:58295","route":"/tour","t":"ui_add"}
2021/04/07 19:41:10 # {"addr":"127.0.0.1:58295","t":"ui_drop"}
2021/04/07 19:41:10 # {"addr":"127.0.0.1:58329","route":"/tour","t":"ui_add"}
2021/04/07 20:09:34 # {"addr":"127.0.0.1:58675","route":"/tour","t":"ui_add"}
2021/04/07 20:09:37 # {"addr":"127.0.0.1:58675","t":"ui_drop"}
2021/04/07 20:18:31 # {"addr":"127.0.0.1:58329","t":"ui_drop"}`

but all time the page shows only the blinking banner "Wave LIVE"

**Reply by lo5** [reply]
@freegor That's odd - you should be seeing an `app_add` in the Wave server's output the moment you start the tour app like in @tkmaker's output:

```
2021/04/07 13:15:53 # {"host":"http://127.0.0.1:8000","route":"/tour","t":"app_add"}
```

If the tour app had failed to connect to the Wave server, you'd see:
```
httpx.ConnectError: [Errno 111] Connect call failed ('127.0.0.1', 10101)
```

In your case, neither happened, which is odd.

**Reply by freegor** [reply]
🤷🏻‍♂️ idk
I try to update the OS to the Big Sur. Don't really think that it may help.
any other suggestions?

**Reply by freegor** [reply]
so have you an idea why the /tour app_add doesn't work for me? Have you guys tied it on the Mac OS Catalina or Big Sur?

**Reply by freegor** [reply]
I thought maybe it related to the VPN connection I have. After I stopped both _wave_ and _tour_ I'm not able to start _tour_ 
It always throws an error (see attached [bash_list.txt](https://github.com/h2oai/wave/files/6280065/bash_list.txt))
and wave runs as expected.

**Reply by tkmaker** [reply]
@freegor I am currently on Big Sur and it works fine for me. 

Can you try changing this line in tour.py to see if using another port for the tour works?
` _app_port = '10102'`

I am wondering if it has to do something about that port on your machine. 

Can you also try just running `hello_world.py` instead of `tour.py`  and then see if the card shows up at http://localhost:10101/demo
`wave run --no-reload examples.hello_world`

**Reply by freegor** [reply]
could you please take a look to the attached output in the message above?

**Reply by lo5** [reply]
From bash_list.txt, it looks like the app is unable to reach the wave server at `http://127.0.0.1:10101` - maybe you have a misconfigured `/etc/hosts`? You can set the ip:port explicitly like this:
```
$ H2O_WAVE_ADDRESS=http://your_ip:your_port wave run --no-reload examples.tour
```

**Reply by freegor** [reply]
_| From bash_list.txt, it looks like the app is unable to reach the wave server at http://127.0.0.1:10101_
But I see the banner in the browser, the wave server is up..
_|  You can set the ip:port explicitly like this:_
Doesn't work, the same output
🥺
Have we maybe some debug log level?

**Reply by freegor** [reply]
so I'm confusing, how can I debug it? First time it started, but stopped working later.
Always it fails on:
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/httpx/_exceptions.py", line 326, in map_exceptions
    yield
.
.

**Reply by freegor** [reply]
ok, if I connected to the VPN, the both server and examples are starting, but I see only logo on the _http://localhost:10101/tour_
if I'm **not connected** to the VPN, the server is running, but examples (or any) don't start.
I have defined proxy and bypass proxy in my .bash_profile:
```
...
export no_proxy=localhost,127.0.0.0/8
export NO_PROXY=${no_proxy}
```
the examples still don't start.
Someone have Wave working with proxy/VPN?

**Reply by lo5** [reply]
Let's try something else:

```
(venv) wave-0.13.0-darwin-amd64 user$  H2O_WAVE_ADDRESS=http://localhost:10101 python examples/hello_world.py
```

After you run that, what do you see at http://localhost:10101/demo?

**Reply by freegor** [reply]
@lo5 I see:
`2021/04/09 20:23:50 # {"address":":10101","t":"listen","webroot":"/Users/user/wave/www"}
2021/04/09 20:25:46 * /demo {"d":[{"k":"hello","d":{"view":"markdown","box":"1 1 2 2","title":"Hello World!","content":"And now for something completely different!"}}]}`
in console and same hello world bla bla text in the browser

**Reply by tkmaker** [reply]
@freegor That is promising. So the hello_world app does work as expected. It sounds like the `tour.py` app which uses port 10102 has an issue possible because of the port it runs on. Did you try the earlier suggestion to change that line in `tour.py` to see if a different port works ? 
It is this line in the `tour.py` file : 
`_app_port = '10102'`

But it looks like standalone apps are working fine. You can try running any of the examples similar to `hello_world.py` in the `examples` folder and navigate to `/demo` to see the app similar to what you did for `hello_world`

**Reply by freegor** [reply]
@tkmaker  I tried to change the `_app_port` to a different value but have the same error.
[bash_list-1.txt](https://github.com/h2oai/wave/files/6288314/bash_list-1.txt)

**Reply by freegor** [reply]
@tkmaker If I run `$ H2O_WAVE_ADDRESS=http://localhost:10101 python examples/tour.py`, I see in the terminal:
_Welcome to the H2O Wave Interactive Tour!
 Go to http://localhost:10101/tour_
But in the browser `http://localhost:10101/tour` I see only Wave Live logo

**Reply by freegor** [reply]
Finally after few days of experiments I have resolved it.
Seems the `export no_proxy=127.0.0.0/8` doesn't work, changed it to `export no_proxy=127.0.0.1` and everything works perfect now. 
Thanks a lot gays all of you for trying help me, I appreciate it!
